[JSC] Implement `Array#toSpliced` in C++

sosukesuzuki opened the pull request:
#### c52c8eeff4dce2712b1420f2b1f6d83a7183aa69
<pre>
[JSC] Implement `Array#toSpliced` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=291870">https://bugs.webkit.org/show_bug.cgi?id=291870</a>

Reviewed by Yusuke Suzuki.

This patch implements `Array#toSpliced` in C++.

                                                  TipOfTree                  Patched

array-prototype-toSpliced-double-double       138.8817+-17.6257    ^    105.4402+-14.7788       ^ definitely 1.3172x faster
array-prototype-toSpliced-int32-int32         149.2978+-14.7220         122.7447+-15.4135         might be 1.2163x faster
array-prototype-toSpliced-int32-contiguous
                                              140.1942+-7.7697     ^    108.4701+-11.8983       ^ definitely 1.2925x faster
array-prototype-toSpliced-double-contiguous
                                              149.8610+-20.7910         115.2025+-19.0198         might be 1.3008x faster
array-prototype-toSpliced-contiguous-contiguous
                                              146.6007+-1.9984     ^    100.1849+-13.2062       ^ definitely 1.4633x faster
array-prototype-toSpliced-int32-double        143.4620+-11.6133         120.8056+-12.1655         might be 1.1875x faster

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::argumentClampedIndexFromStartOrEnd):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: <a href="https://commits.webkit.org/294096@main">https://commits.webkit.org/294096@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f31c276a3f523c26c8769f35ed4cafc91e66c7d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100822 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10773 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105959 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51411 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28948 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76768 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33809 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15976 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91069 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15786 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50787 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93486 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85687 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9152 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108314 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99430 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27940 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85731 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28303 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85274 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21702 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29988 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7718 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21946 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27875 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33132 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123056 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27686 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34295 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29244 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -2294,6 +2295,83 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncCopyWithin, (JSGlobalObject* globalObject
+        return { };
+    }
+
+    uint32_t k = 0;
```
Use `uint64_t`.

Constellation commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -2294,6 +2295,83 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncCopyWithin, (JSGlobalObject* globalObject
+        RETURN_IF_EXCEPTION(scope, { });
+    }
+
+    for (unsigned i = 0; i < insertCount; i++, k++) {
```
Use `uint64_t`.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,106 @@ bool JSArray::fastCopywithin(JSGlobalObject* globalObject, uint64_t from64, uint
+            return nullptr;
+
+        IndexingType insertedItemsIndexingType = sourceType;
+        for (unsigned i = 0; i < insertCount; i++)
```
Use uint64_t.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,106 @@ bool JSArray::fastCopywithin(JSGlobalObject* globalObject, uint64_t from64, uint
+        resultButterfly->setPublicLength(newLength);
+
+        auto copyArrayPrefixElements = [&]<typename T, typename U>(T* resultBuffer, U* sourceBuffer)
+        {
```
We do not put `{` in a new line for lambda.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,106 @@ bool JSArray::fastCopywithin(JSGlobalObject* globalObject, uint64_t from64, uint
+            copyArrayElements<ArrayFillMode::Empty, NeedsGCSafeOps::No>(resultBuffer, 0, sourceBuffer, 0, start, sourceType);
+        };
+        auto copyArraySuffixElements = [&]<typename T, typename U>(T* resultBuffer, U* sourceBuffer)
+        {
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,106 @@ bool JSArray::fastCopywithin(JSGlobalObject* globalObject, uint64_t from64, uint
+                copyArrayPrefixElements(resultBuffer, sourceButterfly->contiguousDouble().data());
+            else
+                copyArrayPrefixElements(resultBuffer, sourceButterfly->contiguous().data());
+            for (unsigned i = 0; i < insertCount; ++i) {
```
Use `uint64_t`.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,106 @@ bool JSArray::fastCopywithin(JSGlobalObject* globalObject, uint64_t from64, uint
+                copyArrayPrefixElements(resultBuffer, sourceButterfly->contiguousDouble().data());
+            else
+                copyArrayPrefixElements(resultBuffer, sourceButterfly->contiguous().data());
+            for (unsigned i = 0; i < insertCount; ++i) {
```
Use `uint64_t`.

Constellation reviewed (changes requested):
Nice. Pointing out some changes.

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -856,6 +856,104 @@ bool JSArray::fastCopyWithin(JSGlobalObject* globalObject, uint64_t from64, uint
+        Butterfly::clearOptimalVectorLengthGap(resultIndexingType, resultButterfly, vectorLength, newLength);
+        JSArray* result = createWithButterfly(vm, nullptr, resultStructure, resultButterfly);
+
+        return result;
```
We should just return immediately.

```
return createWithButterfly(vm, nullptr, resultStructure, resultButterfly);
```

sosukesuzuki committed:
[JSC] Implement `Array#toSpliced` in C++
https://bugs.webkit.org/show_bug.cgi?id=291870

Reviewed by Yusuke Suzuki.

This patch implements `Array#toSpliced` in C++.

                                                  TipOfTree                  Patched

array-prototype-toSpliced-double-double       138.8817+-17.6257    ^    105.4402+-14.7788       ^ definitely 1.3172x faster
array-prototype-toSpliced-int32-int32         149.2978+-14.7220         122.7447+-15.4135         might be 1.2163x faster
array-prototype-toSpliced-int32-contiguous
                                              140.1942+-7.7697     ^    108.4701+-11.8983       ^ definitely 1.2925x faster
array-prototype-toSpliced-double-contiguous
                                              149.8610+-20.7910         115.2025+-19.0198         might be 1.3008x faster
array-prototype-toSpliced-contiguous-contiguous
                                              146.6007+-1.9984     ^    100.1849+-13.2062       ^ definitely 1.4633x faster
array-prototype-toSpliced-int32-double        143.4620+-11.6133         120.8056+-12.1655         might be 1.1875x faster

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::argumentClampedIndexFromStartOrEnd):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:

Canonical link: https://commits.webkit.org/294096@main
